feat(snql): SnQL Grammar and Visitor, Part 2

ayirr7 opened the pull request:
Updates:

- Now supports conditions (WHERE and HAVING clauses) along with previous clauses
- An added test case shows slightly more complicated combination of conditions, with both AND / OR
- Added more type hints, comments

Next steps:
- Any revisions
- Adding parentheses (and priorities) to grammar, visitor

ayirr7 committed:
Added condition logic in grammar, visitor.

ayirr7 committed:
Added condition to test case.

ayirr7 committed:
Add logic for WHERE clause.

ayirr7 committed:
Add another test case for combined conditions.

ayirr7 committed:
Add type hints.

lynnagara commented on snuba/query/snql/__init__.py:
```diff
@@ -29,19 +35,23 @@
 snql_grammar = Grammar(
     fr"""
-    query_exp             = match_clause where_clause? collect_clause? group_by_clause? having_clause? order_by_clause?
+    query_exp             = match_clause where_clause collect_clause group_by_clause? having_clause? order_by_clause?
```
How come the where and collect clauses are now mandatory? Wouldn't a query without these still be valid?

ayirr7 replied:
I think that makes sense - I gave it some thought and compared with Cypher query language as well - it seems to me that a query would still be valid without a WHERE or COLLECT?

lynnagara commented on snuba/query/snql/__init__.py:
```diff
@@ -29,19 +35,23 @@
     order_by_clause       = space* "ORDER BY" order_list space*
 
+    condition             = low_pri_arithmetic condition_op (column_name / numeric_literal)
+    condition_op          = "=" / "!=" / ">" / ">=" / "<" / "<="
```
Should "IN" be on this list as well?

ayirr7 replied:
Currently, no, I'm handling the other conditional operators first. The next step is to deal with parenthetical expressions (those that indicate priority), and if time permits, I would deal with the other operators like "IN".

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -29,19 +35,23 @@
+    having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
 
+    condition             = low_pri_arithmetic condition_op (column_name / numeric_literal)
```
I think you also need string literals on the right hand side of this expression.
Also thinking over this, I think we will need to support functions on the right hand side as well. 
Think about `timestamp > toDateTime('2020-01-01')`
I don't like making the language silently convert the string type in the datetime. We can introduce that conversion later on and limit the allowed functions on the right hand side on after parsing.

A good idea here would be to ask for opinions on the SnQL discussion channel. You could ask people there what they think, if they prefer `timestamp > '2020-01-01'` like we do today or if they prefer to provide the function call explicitly. In the meantime you can accept functions.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -88,13 +97,15 @@ class SnQLVisitor(NodeVisitor):
-        return Query({}, None, collect, None, None, None, groupby, None, orderby)
+        if isinstance(having, Node):
+            having = None
+        return Query({}, None, collect, None, where, None, groupby, having, orderby)
```
please assign these arguments by key, so what is null would be more clear.
```
return Query(
   body={},
   data_source=....
   ...
}

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
 
+    def visit_where_clause(
+        self, node: Node, visited_children: Tuple[Any, Any, Sequence[Expression], Any]
+    ) -> Sequence[Expression]:
```
I think this should be returning a single Expression that represent the whole boolean conditions tree not a sequence of expressions.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
+
+    def visit_having_clause(
+        self, node: Node, visited_children: Tuple[Any, Any, Sequence[Expression], Any]
+    ) -> Sequence[Expression]:
```
Same as above:
I think this should be returning a single Expression that represent the whole boolean conditions tree not a sequence of expressions.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
+        # in the case of one Condition
+        # and_condition will be an empty Node
+        if not isinstance(and_condition, Node):
+            _, right_condition = and_condition
```
Isn't `and_condition` a sequence of multiple `["AND", Expression]` ? So shouldn't you iterate over right_condition ?
If not, I am not sure I get who is combining the the multiple conditions in AND.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
+        # or_condition will be an empty Node
+        if not isinstance(or_condition, Node):
+            _, right_condition = or_condition
+            return combine_or_conditions([left_condition, right_condition])
```
same as above

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -151,6 +162,55 @@ def visit_quoted_literal(
+            return combine_or_conditions([left_condition, right_condition])
+        return left_condition
+
+    def visit_condition_comma(
+        self, node: Node, visited_children: Tuple[Expression, Any, Any, Any]
+    ) -> Expression:
+        bin_condition, _, _, _ = visited_children
+        return bin_condition
```
What is this for ?

ayirr7 replied:
Removed.

fpacifici commented on tests/query/snql/test_query.py:
```diff
@@ -31,16 +36,84 @@
+            ),
         ),
-        id="Simple COLLECT, GROUPBY, ORDERBY clause example",
+        id="Example 2",
```
I would suggest to add a few simpler tests that test some specific conditions:
1) multiple conditions in AND (more than 2). The other clauses can be trivial
2) multiple conditions in OR (more than 2). The other clauses can be trivial
3) Conditions with different types of literals (int, columns, strings)
4) Combination of AND and OR at multiple levels.

This approach will allow you to cover more cases without having a massive query each time to express with the AST.

fpacifici reviewed (changes requested):
Seems a good step in the right direction, please see the comments inline.

ayirr7 committed:
Added condition test cases.

ayirr7 committed:
Modified logic for conditions, improved type hints and grammar rules.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitorp2

ayirr7 commented:
Updates:

- Modified the logic for dealing with conditions in Visitor
- Added function_call to right hand side of a condition in grammar
- Added test cases for various conditions/combinations of conditions
- Improved some type hints

Next Steps:

- PR for parentheses

ayirr7 committed:
Made WHERE and COLLECT clauses optional.

ayirr7 committed:
Cosmetic changes.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -29,37 +34,46 @@
+    having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
 
+    condition             = low_pri_arithmetic condition_op (function_call / column_name / numeric_literal) space*
```
I think you are missing the string literal.

ayirr7 replied:
Added.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -29,37 +34,46 @@
 
+    condition             = low_pri_arithmetic condition_op (function_call / column_name / numeric_literal) space*
+    condition_op          = "=" / "!=" / ">" / ">=" / "<" / "<="
+    interm_condition      = open_paren or_expression close_paren
```
let's give this a better name. Like `parenthesized_condition` ?

ayirr7 replied:
Will modify names in P3 (where the parenthesized conditions will be).

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -29,37 +34,46 @@
 
     low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
-    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    high_pri_arithmetic   = (space* arithmetic_term space* (high_pri_tuple)*)
```
why did you add parenthes here ?

ayirr7 replied:
Removed.

fpacifici commented on tests/query/snql/test_query.py:
```diff
@@ -31,16 +32,254 @@
-                OrderBy(OrderByDirection.ASC, Column(None, None, "m")),
+            order_by=[OrderBy(OrderByDirection.DESC, Column(None, None, "f"))],
+        ),
+        id="Example 1",
```
Please provide a descriptive name to the test. It should convey what this test is for.

ayirr7 replied:
Added.

fpacifici reviewed (approved):
Please address the comments before merging.

evanh commented on tests/query/snql/test_query.py:
```diff
@@ -2,14 +2,15 @@
 from snuba.query.snql import parse_snql_query
 
 
 test_cases = [
```
I think you should also start adding some negative tests, i.e. queries that shouldn't work.

For example, I'm curious what happens if I omit spaces from my query. Based on the regex, spaces seem to be totally optional.

Is `MATCH(blah)WHEREa<3COLLECT4-5` a valid query?

ayirr7 replied:
Hi, I've started adding some test cases that serve as invalid queries that hopefully begin to illustrate what sorts of queries cannot be parsed and why. Would appreciate any further suggestions to add more cases!

ayirr7 committed:
Revisions.

ayirr7 committed:
Code revisions.

ayirr7 committed:
Added more test cases to illustrate parser behavior.

ayirr7 committed:
Added negative tests.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitorp2

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -29,37 +34,45 @@
-    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+
+    arithmetic_term       = space* (function_call / numeric_literal / column_name / interm_arithm) space*
+    interm_arithm         = open_paren low_pri_arithmetic close_paren
```
please give to this one a descriptive name.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -139,6 +192,7 @@ def visit_high_pri_arithmetic(
         node: Node,
         visited_children: Tuple[Any, Expression, Any, HighPriArithmetic],
     ) -> Expression:
+
```
please revert this change.

fpacifici commented on tests/query/snql/test_invalid_queries.py:
```diff
@@ -0,0 +1,32 @@
+        id="ORDER BY is two words",
+    ),
+    pytest.param(
+        "MATCH(blah)WHEREa<3COLLECT4-5,3*g(c),cBYd,2+7ORDER BYfDESC",
```
mind adding the spaces between match, where and collect, if you are testing the ORDER BY, the rest of the query can still be valid.

fpacifici commented on tests/query/snql/test_invalid_queries.py:
```diff
@@ -0,0 +1,32 @@
+        id="Expression before ASC / DESC needs to be separated from ASC / DESC keyword by space",
+    ),
+    pytest.param(
+        "MATCH(blah)WHEREa<3COLLECT4-5,3*g(c),cBYd,,2+7 ORDER BY f DESC",
```
same comment.

ayirr7 committed:
Minor cosmetic changes.

ayirr7 committed:
Added spaces to test cases.
